### ISSUE TITLE
Handling of repositories with submodules

### ISSUE THREAD

**Issue by jpschewe** [FredrikNoren/ungit#1013]
It seems that the new submodules structure confuses mungit. With recent versions of git the .git directory for submodules is inside the .git directory of the parent repository.

```
2017-12-12T15:15:50.491Z - warn: Responding with ERROR:  {"errorCode":"not-a-repository","error":"No such file: /Users/jschewe/projects/map/wd.map-code/src/P2Protelis/.git/HEAD"}
```

See discussion at https://github.com/exsilium/mungit/issues/10

**Comment by ylecuyer** [comment]
As of 1.4.15 I'm not able to reproduce.

```
git clone git@github.com:jpschewe/rootsh.git
cd rootsh
ungit
```

Then go to 

```
http://localhost:8448/#/repository?path=%2Fhome%2Fylecuyer%2FProjects%2Frootsh/continuous_integration/lcov-to-cobertura-xml
```

No problems

You should try again and close your issue if it's working as expected @jpschewe

**Comment by jpschewe** [comment]
I just tried again with commit 58a1c03ff61ff68298cd80b0b0e94956995a5fe2 and I still get the errors. They are listed below.
```
2018-03-23T12:38:44.560Z - warn: Responding with ERROR:  {"errorCode":"not-a-repository","error":"No such file: /home/jpschewe/projects/rootsh/working-dir/continuous_integration/lcov-to-cobertura-xml/.git/HEAD"}
2018-03-23T12:38:45.648Z - warn: Responding with ERROR:  {"errorCode":"not-a-repository","error":"No such file: /home/jpschewe/projects/rootsh/working-dir/continuous_integration/lcov-to-cobertura-xml/.git/HEAD"}
```

When you test this which version of git are you using?
I have 2.14.1.
 
When you look at continuous-integration/lvoc-to-cobertura-xml on the filesystem, is `.git` a directory or a file? For me it's a file that points to a different git directory: 
```
[git:master] >cat .git
gitdir: /home/jpschewe/projects/rootsh/working-dir/.git/modules/continuous_integration/lcov-to-cobertura-xml
```

**Comment by jung-kim** [comment]
Can you do me a favor and share with me the content of `.gitmodules` file in the troubled root git repo?  

`.git` is the folder where git states and metadata is stored.  

Quick test on my side works and I think either it is a mis configuration or ungit is not handling it correctly.

**Comment by jpschewe** [comment]
The .gitmodules is in the repository.
This behavior seems to be the default in new versions of git. See the absorbgitdirs option to the git submodule command.

**Comment by jung-kim** [comment]
Hmmm...  It's a file for me. and my git version is 2.14.3.  I may have to setup a different env with test with different git versions later.  

<img width="409" alt="screen shot 2018-03-26 at 20 42 00" src="https://user-images.githubusercontent.com/5281068/37945691-3eefa4ae-3136-11e8-88e9-0a6096222d21.png">

**Comment by jpschewe** [comment]
I don't think it has anything to do with the .gitmodules file. It has to do with the fact that .git in the submodule is a file instead of a directory and ungit doesn't appear to handle that properly.

**Comment by jung-kim** [comment]
What is the submodule path in above failure?

**Comment by jpschewe** [comment]
@jung-kim I don't understand what you are asking.

**Comment by jung-kim** [comment]
in `2017-12-12T15:15:50.491Z - warn: Responding with ERROR:  {"errorCode":"not-a-repository","error":"No such file: /Users/jschewe/projects/map/wd.map-code/src/P2Protelis/.git/HEAD"}`.

is `/Users/jschewe/projects/map/wd.map-code/src/P2Protelis/` a submodule?

**Comment by jpschewe** [comment]
That's from another repository with the same issue. The error from this repository is a few comments back.
```
2018-03-23T12:38:44.560Z - warn: Responding with ERROR:  {"errorCode":"not-a-repository","error":"No such file: /home/jpschewe/projects/rootsh/working-dir/continuous_integration/lcov-to-cobertura-xml/.git/HEAD"}
```

**Comment by jpschewe** [comment]
It's not specific to the repository, it seems to be tied to the version of git using to clone the repository. See discussion at https://github.com/exsilium/mungit/issues/10.

**Comment by jpschewe** [comment]
I did some poking into the source on this. It seems that ungit assumes that if it finds something in the filesystem named '.git' that it's a directory and that the git things can be found in it. It turns out this isn't always true as '.git' can be file file with "gitdir: <path>" inside it telling git to look somewhere else for the git directory. I expect that ungit needs to be refactored some to have a method that finds the git directory and this method called wherever the git directory is needed. This method will find a filesystem entry that is ".git" and if it's a file, parse it, otherwise use it like it currently does.

**Comment by jung-kim** [comment]
Yikes, ungit does assumes that all valid git repository location has `.git` folder, and if this is no longer true this does complicates a lot as we would have to manage two different git repo types.  

I've just got back from a vacation so I didn't get to study about this a lot but I'm having hard time replicating this `.git` file.  I've upgraded to `git@2.15.1` and wasn't able to create the `.git` file via init or clone.  

can you share the steps you did so I can recreate and test?

**Comment by jpschewe** [comment]
@jung-kim as stated above, I just clone the repository to make it happen: `git clone git@github.com:jpschewe/rootsh.git`. 

I just verified that this happens on both git 2.11.0 and 2.14.1 under Linux. If this is not happening for you, then I suggest that you look at `git submodule absorbgitdirs` to convert an existing repository.

From the help for git submodule:
```
       absorbgitdirs
           If a git directory of a submodule is inside the submodule, move the git directory of the
           submodule into its superprojects $GIT_DIR/modules path and then connect the git
           directory and its working directory by setting the core.worktree and adding a .git file
           pointing to the git directory embedded in the superprojects git directory.

           A repository that was cloned independently and later added as a submodule or old setups
           have the submodules git directory inside the submodule instead of embedded into the
           superprojects git directory.

           This command is recursive by default.
```

**Comment by jung-kim** [comment]
Was able to replicate and conduct tests.  Here are some pointers.

* `absorbgitdirs` only applies to submodules inside of a submodule that is in git.  So it doesn't apply in this case and not applicable to `lcov-to-cobertura-xml` submodule at  `https://github.com/jpschewe/rootsh/`
* This `not-a-repository` warning is is from the [/checkout](https://github.com/FredrikNoren/ungit/blob/master/source/git-api.js#L469) API call when a user navigate into the submodule from Ungit UI.  
  * This submodule is added via git hash, [`1be40a82`](https://github.com/eriwen/lcov-to-cobertura-xml/tree/1be40a820f4b7ec5cf68251bb4eacf720b23e219) not by branch or tag.  Ungit requires either branch or tag as some of the operations becomes difficult or doesn't make sense
  * Navigation into a submodule from Ungit's UI is for reference only.  One shouldn't make any git operations against submodule.  If one wish to do so, one should do so against the original git repo of the submodule and update submodule once those changes are submitted.  (currently Ungit allows some git operations and I may make  navigation into a submodule to a read only view)

This warning is not an error and requires a fix per say.  I would say we need to make navigation into a submodule from Ungit UI needs to be read only with notifications and disabled git operations.

@jpschewe let me know if I missed something.

**Comment by jpschewe** [comment]
Yes absorbgitdirs only applies to submodules. However git does allow a developer to move the .git directory in this way for any git repository. See https://git-scm.com/docs/gitrepository-layout
and https://stackoverflow.com/questions/505467/can-i-store-the-git-folder-outside-the-files-i-want-tracked

Yes, this submodule is checked out via git hash, but this problem occurs even if the submodule is checked out by branch or tag.

You may not make changes in git submodules directly, but many other developers do.

It's an error because I can't make changes or even properly view the state of the submodule because the errors keep coming up.

Here's pseudocode for a function that would solve the problem.
```js
resolveGitDir(gitDirOrFile) {
  if(isDir(gitDirOrFile)) {
    return gitDirOrFile
  } else {
    firstLine = gitDirOrFile.read()
    regexp = /^gitdir:\s*\(.*)$/
    match = regexp.exec(gitDirOrFile)
    if(match) {
      gitDir = match[0]
      if(gitDir[0] = '"' && gitDir[gitDir.length - 1] = '"') {
        return gitDir.substring(1, gitDir.length - 1)
      } else if(gitDir[0] = "'" && gitDir[gitDir.length - 1] = "'") {
        return gitDir.substring(1, gitDir.length - 1)
      } else {
        return gitDir
      }
    } else {
      error
    }
}
```

Then wherever you reference ".git" in the code, just call resolveGitDir(".git") and everything should work. You might want to cache the result of this if the function is slow, although then there is complexity in making sure to clear the cache when the current directory changes.

**Comment by jung-kim** [comment]
I would have to find the submodule with branch that is a `.git` file example to understand it's content and how to parse them.

As far as viewing submodule, I didn't see any errors on ungit side when viewing the submodule and had no problem with navigation.  how are you getting these errors?

As far as changing submodule, you shouldn't change submodule.  You should change the module and update the submodule later.

Properly handing `.git` file vs. folder is not a simple problem as what is being suggested.  There are several file watches on files of `.git` that needs to be reworked, file parsing will be different.

**Comment by jpschewe** [comment]
I gave you pseudocode for the parse function and have pointed you at the git documentation. 

Like I said in the first post, I'm seeing these errors when viewing the submodule repository from inside ungit. However you won't see them unless you have moved the `.git` directory.

Anywhere you reference `.git`, just reference the function. That will return the correct path. This includes creating the file watches.

**Comment by jung-kim** [comment]
Your first post is a warning about a git error.  Ungit should be functional.

I understand you see them in console, but does it block you from viewing the repo? it shouldn't block you from viewing the submodule.

It seems that you are more passionate about this issue, feel free to send PRs or clone the repo

**Comment by jpschewe** [comment]
The error in the first post is from ungit not handling this file.

I can sometimes view the repo, but it doesn't update or allow me to change branches.

**Comment by jpschewe** [comment]
I will see about a PR, I'm not a NodeJS developer, so I'm hoping that someone that knows the code and language better will be able to take care of this bug.

**Comment by jung-kim** [comment]
You are absolutely right that ungit is not handling that file correctly. Because of that git error is thrown and ungit is dismiss that git error as a warning. 

Not being able to update is a problem and yes I think that should be fixed.  But you should be able to update submodule at the parent git directory via submodule update method. 

You shouldn't create branches in submodules.  Should you want to change submodule it should be done from parent git directory via git commands that are not yet supported via ungit.  This is because it will potentially create mismatches between module git metadata at the parent directory and the actual submodule state.  

-----
I'm going to try to fix to things.  Make submodule view read only with proper warning sings, and try to fix the git fetch functionality when within git submodule.

**Comment by jpschewe** [comment]
Making submodules read-only would make ungit unusable for me and likely others that have tightly coupled submodules where one needs to make changes to them and the parent git repository at the same time.

**Comment by jpschewe** [comment]
I'm not sure where you get this idea that one should not make changes in a submodule.

**Comment by jung-kim** [comment]
https://stackoverflow.com/questions/5427414/editing-a-git-submodule

```
Bottomline:

Yes. Developing within a submodule folder is possible and often convenient but not without its risks. Choose your development model wisely
```

**Comment by jpschewe** [comment]
I have decided to go with @jung-kim suggestion of not developing inside my submodules. So I won't be working on this merge request. Others may be interested in it if they want to use the alternate git directory functionality in git.

**Comment by mtryan83** [comment]
Unfortunately, I think this issue needs to be revisited since it's now completely crashing ungit. 

Going by the second two answers [here](https://stackoverflow.com/questions/64122687/is-a-git-submodule-an-independent-repository), the default for submodules going forward appears to be the file version of .git (for probably a good reason, see e.g. bk2204's answer), regardless of whether the submodule is read-only/changed/etc. 

So for example we can reproduce the problem with:
(using git version 2.34.1 (latest) and the test-submodules repository from @jpschewe)
`git clone git@github.com:jpschewe/test-submodules.git`
`cd test-submodules`
`git rm -r submodules-test #submodules-test is a regular directory in the repo`
`git commit -m "removed submodules-test regular directory"`

From here we have two choices: adding the submodule from the command-line 
`git submodule add git@github.com:jpschewe/submodule-test.git`
or using the "add submodule" choice under the submodules menu. Either choice does the same thing

Commit the submodule
(either `git commit -m "added submodule from command line"` or from ungit)

Then if I try to visit the submodule (either through the link in the submodule heading or just directly through the address bar), ungit _completely_ crashes with the error
`2022-01-25T20:16:10.148Z - error: Error: ENOTDIR: not a directory, mkdir '/home/XXXXX/workspace/test-submodules/submodule-test/.git/HEAD'`

(Note there may be some warnings about submodule-test already existing in the origin. This is not relevant, the same ENOTDIR error occurs when using other repos as submodules)

I don't know much about NodeJS, so I don't know how to help with the pr.
